### DISCUSSION TITLE
Reason for unique PID for different device configurations

### DISCUSSION THREAD

**Discussion by Jubatian** [hathach/tinyusb#2022] [Q&A] [answered]
Hi,

Would like to ask about a remark I found in the codebase, here:

https://github.com/hathach/tinyusb/blob/master/examples/device/midi_test/src/usb_descriptors.c#L28

"A combination of interfaces must have a unique product id, since PC will save device driver after the first plug. Same VID/PID with different interface e.g MSC (first), then CDC (later) will possibly cause system error on PC."

What is the reason behind this? I am working on a project which has a USB device component which can be turned On/Off by the user (effect taking place after device reboot), and I experience a strong push to keep doing this on a single PID. So if any adverse effect is known, might like to see it, hopefully to steer us towards a better decision.

**Comment by hathach** [comment, marked as answer]
the reason is obvious, host may cache driver installation with vid/pid, it may or may not re-discover the driver for device everytime it enumerates to save time. In correct driver call e.g MSC driver on HID endpoint may or may lead to kernel error especially with driver that requires inf file. In my experience, windows 7 got to BSOD easily with this error, later version of windows may catch this. But I don't use windows for a very long time, thing may be improved now.

It may be safer if your device does not need an inf file to work, but make sure you test this thoroughly on all platforms

**Reply by Jubatian** [reply]
Awesome, thank you!

Seems like we may have dodged most of this earlier as we used constant endpoint numbers (with a different USB stack), so no situations could arise when a different thing would appear on the endpoint (only the endpoint would be missing if that device component was turned off).
